### PR TITLE
feat(core): add gradle plugin

### PR DESCRIPTION
**Pull request by xiongemi**
## Related Issue(s)

Fixes #

### CODE REVIEW

File: .github/workflows/e2e-matrix.yml [hunk 1]
```diff
@@ -112,6 +112,7 @@ jobs:
           - e2e-esbuild
           - e2e-eslint
           - e2e-expo
+          - e2e-gradle
```
Comment (meeroslav) [inline .github/workflows/e2e-matrix.yml:115]: What about windows nightly? Do we not care about that one or it doesn't work?

**Review by meeroslav — APPROVED** [review]

File: .github/workflows/e2e-matrix.yml (hunk 1 above)
Comment (xiongemi) [reply .github/workflows/e2e-matrix.yml]: removed

**Review by mandarini — APPROVED** [review]

File: packages/gradle/src/plugin/dependencies.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+          depsFile,
+          gradleProjectToNxProjectMap,
+          source,
+          gradleFile,
+          context
+        )
+      );
+    }
+  }
+  console.timeEnd('locating gradle dependencies');
```
Comment (AgentEnder) [inline packages/gradle/src/plugin/dependencies.ts:60]: Can we change this to performance.mark / performance.measure so they show up with NX_PERF_LOGGING?

File: packages/nx/src/project-graph/project-graph-builder.ts [hunk 3]
```diff
@@ -383,11 +383,13 @@ export class ProjectGraphBuilder {
...
-            fileDeps.set(target, new Set([fileDataDepType(d)]));
-          } else {
-            fileDeps.get(target).add(fileDataDepType(d));
+          if (d) {
+            const target = fileDataDepTarget(d);
+            if (!fileDeps.has(target)) {
+              fileDeps.set(target, new Set([fileDataDepType(d)]));
+            } else {
+              fileDeps.get(target).add(fileDataDepType(d));
+            }
```
Comment (AgentEnder) [inline packages/nx/src/project-graph/project-graph-builder.ts:392]: Why was this change needed? I'd not expect d to be falsy.

File: packages/gradle/src/plugin/dependencies.ts (hunk 2 above)
Comment (xiongemi) [reply packages/gradle/src/plugin/dependencies.ts]: changed.

File: packages/nx/src/project-graph/project-graph-builder.ts (hunk 3 above)
Comment (xiongemi) [reply packages/nx/src/project-graph/project-graph-builder.ts]: i think i ran into an issue (do not remember exactly). i removed it now and it seems fine

File: packages/gradle/src/plugin/dependencies.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  }
+  const gradleDependenciesEnd = performance.mark('gradleDependencies:end');
+  performance.measure(
+    'gradleDependencies',
+    gradleDependenciesStart.name,
+    gradleDependenciesEnd.name
+  );
+  return dependencies;
+};
+const gradleConfigFileNames = new Set(['build.gradle', 'build.gradle.kts']);
```
Comment (AgentEnder) [inline packages/gradle/src/plugin/dependencies.ts:68]:
```suggestion

const gradleConfigFileNames = new Set(['build.gradle', 'build.gradle.kts']);
```

File: packages/gradle/src/plugin/nodes.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        return {
+          type: tasksTypeMap.get(taskName) ?? 'Unknown',
+          name: taskName,
+        };
+      });
+
+      const outputDirs = gradleFileToOutputDirsMap.get(gradleFilePath) as Map<
+        string,
+        string
+      >;
```
Comment (AgentEnder) [inline packages/gradle/src/plugin/nodes.ts:102]: All of this should be done only if there isn't a cache hit, otherwise its doing work that is already present in cache.

File: packages/gradle/src/utils/exec-gradle.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  args: string[],
+  execOptions: ExecFileSyncOptionsWithBufferEncoding
+) {
+  const gradleBinaryPath = getGradleBinaryPath();
+
+  return execFileSync(gradleBinaryPath, args, execOptions);
+}
+
+export function getGradleBinaryPath() {
+  const gradleBinaryPath = join(workspaceRoot, './gradlew');
```
Comment (AgentEnder) [inline packages/gradle/src/utils/exec-gradle.ts:21]: Make sure this works on windows, I think there it should be `./gradlew.bat`?

File: packages/gradle/src/utils/exec-gradle.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  return gradleBinaryPath;
+}
+
+export function execGradleAsync(
+  args: ReadonlyArray<string>,
+  execOptions: ExecFileOptions
+) {
+  const gradleBinaryPath = join(workspaceRoot, './gradlew');
```
Comment (AgentEnder) [inline packages/gradle/src/utils/exec-gradle.ts:33]: Use `getGradleBinaryPath`

File: packages/gradle/src/utils/get-gradle-report.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+import { workspaceRoot } from '@nx/devkit';
+
+import { execGradle } from './exec-gradle';
+
+let gradleReport: ReturnType<typeof processProjectReports>;
+export function getGradleReport() {
+  if (gradleReport) {
+    return gradleReport;
+  }
```
Comment (AgentEnder) [inline packages/gradle/src/utils/get-gradle-report.ts:12]: This cache may live too long when daemon is active, can you validate that it reflects appropriate changes when a file is updated?

**Review by AgentEnder — CHANGES_REQUESTED** [review]
Looks good. Few comments, nothing way too major.

File: CODEOWNERS [hunk 9]
```diff
@@ -144,6 +144,10 @@ rust-toolchain @nrwl/nx-native-reviewers
 /packages/devkit/public-api.ts @FrozenPandaz @vsavkin
 /packages/devkit/nx.ts @FrozenPandaz @vsavkin
 
+## Gradle
```
[the remark this answers is not in the document]
Comment (xiongemi) [reply CODEOWNERS:147]: changed.

File: packages/gradle/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<p style="text-align: center;"><img src="https://raw.githubusercontent.com/nrwl/nx/master/images/nx.png" width="600" alt="Nx - Smart, Fast and Extensible Build System"></p>
```
[the remark this answers is not in the document]
Comment (xiongemi) [reply packages/gradle/README.md:1]: changed

File: packages/gradle/src/plugin/dependencies.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
+import {
+  CreateDependenciesContext,
+  DependencyType,
+  RawProjectGraphDependency,
+  validateDependency,
+} from '@nx/devkit';
+import { requireNx } from '@nx/devkit/nx';
```
[the remark this answers is not in the document]
Comment (xiongemi) [reply packages/gradle/src/plugin/dependencies.ts:7]: changed

File: packages/gradle/src/plugin/dependencies.ts (hunk 4 above)
Comment (xiongemi) [reply packages/gradle/src/plugin/dependencies.ts:68]: changed

File: packages/gradle/src/utils/exec-gradle.ts (hunk 6 above)
Comment (xiongemi) [reply packages/gradle/src/utils/exec-gradle.ts:21]: changed

File: packages/gradle/src/utils/exec-gradle.ts (hunk 7 above)
Comment (xiongemi) [reply packages/gradle/src/utils/exec-gradle.ts:33]: changed

File: packages/gradle/src/plugin/nodes.ts (hunk 5 above)
Comment (xiongemi) [reply packages/gradle/src/plugin/nodes.ts:102]: changed

File: packages/gradle/package.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  "author": "Victor Savkin",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/nrwl/nx/issues"
+  },
+  "homepage": "https://nx.dev",
+  "ng-update": {
+    "requirements": {},
+    "migrations": "./migrations.json"
+  },
```
[the remark this answers is not in the document]
Comment (xiongemi) [reply packages/gradle/package.json:28]: changed

File: packages/gradle/project.json [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  "sourceRoot": "packages/gradle/src",
+  "projectType": "library",
+  "targets": {
+    "nx-release-publish": {
+      "dependsOn": ["^nx-release-publish"],
+      "executor": "@nx/js:release-publish",
+      "options": {
+        "packageRoot": "build/packages/gradle"
+      }
+    },
```
[the remark this answers is not in the document]
Comment (xiongemi) [reply packages/gradle/project.json:13]: this is needed because Gradle is currently private. this target is only added for non-private repo, however, i still need to publish it for e2e tests:

in packages/nx/src/utils/package-json.ts:
```
  if (!packageJson.private && !res['nx-release-publish']) {
    res['nx-release-publish'] = {
      dependsOn: ['^nx-release-publish'],
      executor: '@nx/js:release-publish',
      options: {},
    };
  }
```

File: e2e/gradle/src/gradle.test.ts
[the remark this answers is not in the document]
Comment (xiongemi) [reply e2e/gradle/src/gradle.test.ts:1]: changed

File: packages/gradle/src/utils/get-gradle-report.ts (hunk 8 above)
Comment (xiongemi) [reply packages/gradle/src/utils/get-gradle-report.ts]: add a function to delete cache

**Review by AgentEnder — APPROVED** [review]
